client/client.go

```diff
@@ -131,7 +133,10 @@ type Client struct {
 	negotiateVersion bool
 
 	// negotiated indicates that API version negotiation took place
-	negotiated bool
+	negotiated atomic.Bool
+
+	// versionLock protects the version and negotiated fields from concurrent writes
+	versionLock sync.Mutex
```

Comment:
This mutex does more than just protect struct fields. (Besides, the `negotiated` field is an atomic; it can be concurrently written to by multiple goroutines without any data races. And since it gets latched to `true`, there would not be a logical race, either.)
It is more accurate to say that this mutex is used to single-flight the version negotiation process. `negotiateLock` was a good name for that; thoughts on renaming it back?